spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/GcpSecretManagerTemplateConfiguration.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright 2017-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.autoconfigure.secretmanager;
+
+import java.io.IOException;
+
+import com.google.api.gax.core.CredentialsProvider;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceClient;
+import com.google.cloud.secretmanager.v1beta1.SecretManagerServiceSettings;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.core.DefaultCredentialsProvider;
+import org.springframework.cloud.gcp.core.DefaultGcpProjectIdProvider;
+import org.springframework.cloud.gcp.core.GcpProjectIdProvider;
+import org.springframework.cloud.gcp.core.UserAgentHeaderProvider;
+import org.springframework.cloud.gcp.secretmanager.SecretManagerTemplate;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+/**
+ * Autoconfiguration for GCP Secret Manager which provides an instance of the
+ * {@link SecretManagerTemplate}.
+ *
+ * @author Daniel Zou
+ * @since 1.2.2
+ */
+@Configuration
+@EnableConfigurationProperties(GcpSecretManagerProperties.class)
+@ConditionalOnClass(SecretManagerServiceClient.class)
+@ConditionalOnProperty(value = "spring.cloud.gcp.secretmanager.template.enabled", matchIfMissing = true)
+public class GcpSecretManagerTemplateConfiguration {
```

Comment:
I'm having second thoughts on the separation of the template from the bootstrap configuration. The main reason being that we wanted to re-use the template in the property source implementation, and currently we don't. If we do create the `SecretManagerTemplate` in the boostrap, does it remain accessible to the application context? If so, I'm not sure of what the benefit would be of this separate configuration class.

Reply:
Also, do we want to support the use-case where someone doesn't want to use bootstrap properties but wants to use the template?
Since loading secrets into the application context inadvertently may be risky, maybe we should even disable it by default, and the only thing you get with the starter by default, would be the template.

Reply:
Right, I created two separate configurations in order to support the use-case where someone doesn't want to use bootstrap properties but wants to use the template.

If you want the template in the property source impl, I can add an additional `@ConditionalOnMissingBean SecretManagerTemplate` to the bootstrap config. Then disable the bootstrap config by default?

Reply:
Resolved offline; see https://github.com/spring-cloud/spring-cloud-gcp/pull/2223#issuecomment-593570686